src/main/java/org/usfirst/frc4048/common/apriltags/TCPServer.java

```diff
@@ -0,0 +1,120 @@
+package org.usfirst.frc4048.common.apriltags;
+
+import edu.wpi.first.wpilibj.DriverStation;
+
+import java.io.BufferedInputStream;
+import java.io.DataInputStream;
+import java.io.IOException;
+import java.net.ServerSocket;
+import java.net.Socket;
+import java.util.LinkedList;
+import java.util.Queue;
+import java.util.concurrent.LinkedTransferQueue;
+
+public abstract class TCPServer<T> extends Thread {
+
+    private final int port;
+    private ServerSocket serverSocket;
+    private Socket clientSocket;
+    private DataInputStream dataInputStream;
+    private BufferedInputStream bufferedInputStream;
+    private boolean running = false;
+    private final LinkedTransferQueue<T> readings = new LinkedTransferQueue<>();
+
+    /**
+     * @param port for network communication. Port must match between client and server (range 5800-5810 (rio limitations))
+     */
+    public TCPServer(int port) {
+        this.port = port;
+    }
+
+
+    @Override
+    public synchronized void start() {
+        try {
+            this.serverSocket = new ServerSocket(port);
+            this.clientSocket = null;
+        } catch (IOException e2) {
+            DriverStation.reportError("Could not start server on port " + port, true);
+            this.serverSocket = null;
+        }
+        running = true;
+        super.start();
+    }
+
+    @Override
+    public void run() {
+        while (running) {
+            if (serverSocket == null) {
+                exit();
+                return;
+            }
+            if (this.clientSocket == null) {
+                clientSocket = waitForClient();
+                if (clientSocket == null) {
+                    continue;
+                }
+                try {
+                    bufferedInputStream = new BufferedInputStream(clientSocket.getInputStream());
+                    dataInputStream = new DataInputStream(bufferedInputStream);
+                } catch (IOException e) {
+                    DriverStation.reportError("Could not fetch client Input Stream", true);
+                    dataInputStream = null;
+                    continue;
+                }
+            }
+            try {
+                readings.add(extractFromStream(dataInputStream));
+            } catch (IOException e) {
+                DriverStation.reportError("Problem while reading tcp stream data", true);
+                clientSocket = null;
+                bufferedInputStream = null;
+                dataInputStream = null;
+            }
+        }
+
+    }
+
+    protected abstract T extractFromStream(DataInputStream stream) throws IOException;
+
+    private Socket waitForClient() {
+        Socket s;
+        try {
+            s = serverSocket.accept();
+        } catch (IOException e1) {
+            DriverStation.reportError("Could not connect to client on port " + port, true);
+            s = null;
+        }
+        return s;
+    }
+
+    public void exit() {
```

Comment:
This actually overrides the super's `exit`. Was that the intent?

Reply:
I did not realize thread had an `exit` method. Overriding was not the intent. I just checked and the thread `exit` method is private, so the signatures would be different, and I don't think it would override the method (Not sure though). Either way, I can change the method name so it is not confusing.

Reply:
Yes please

Reply:
This is just confusing since you get no error for overriding a private method, though the behavior is somewhat expected: Base class will invoke the private method and subclass - the public one.

Reply:
Changed to `escape();`